### PR TITLE
Add linux support

### PR DESCRIPTION
**Pull request by Alph4rd**

### CODE REVIEW

**Comment by dm94** [comment]
I am completely unfamiliar with how it works.
But wouldn't it be possible to detect which operating system is being used and on that basis activate or deactivate functions?
I know that many offsets would have to be changed, but based on a global variable it could be set to use one or the other.

**Comment by fabio1999ita** [comment]
> I am completely unfamiliar with how it works. But wouldn't it be possible to detect which operating system is being used and on that basis activate or deactivate functions? I know that many offsets would have to be changed, but based on a global variable it could be set to use one or the other.

yeah, should be also easy and need to adapt the download of API too

**Comment by Pablete1234** [comment]
> But wouldn't it be possible to detect which operating system is being used and on that basis activate or deactivate functions? I know that many offsets would have to be changed, but based on a global variable it could be set to use one or the other.

Yes, when this PR is to be merged, it will be with such capability to adapt automatically to the OS, atm, the changes in here have been slowly being refactored/merged into master to make this PR more lightweight (less OS-specific things), and when it gets to be a minimum set of differences, those differences will have to work off of a toggle based on the OS, which darkbot will handle automatically.

Regarding browser native APIs, the libraries for this are not redistributed alongside the rest of windows-based apis (yet) because this isn't merged (yet). When this gets official support and is merged into master, the main libs will simply download the correct natives for linux which will be diff to the ones from windows.

File: src/main/java/com/github/manolo8/darkbot/config/Config.java [hunk 1]
```diff
@@ -224,7 +224,7 @@ public static class WindowPosition {
 
         public @Option @Visibility(Level.ADVANCED) APIConfig API_CONFIG = new APIConfig();
         public static class APIConfig {
-            public @Option @Dropdown BrowserApi BROWSER_API = BrowserApi.KEKKA_PLAYER;
+            public @Option @Dropdown BrowserApi BROWSER_API = BrowserApi.TANOS_API;
```
Comment (Pablete1234) [inline src/main/java/com/github/manolo8/darkbot/config/Config.java:227]:
```suggestion
            public @Option @Dropdown BrowserApi BROWSER_API = OSUtils.isLinux() ? BrowserApi.TANOS_API : BrowserApi.KEKKA_PLAYER;
```

File: src/main/java/com/github/manolo8/darkbot/core/manager/MapManager.java [hunk 2]
```diff
@@ -421,7 +421,9 @@ private Location getEnemyLocatorTarget() {
     }
 
     private boolean findMarker(long spriteArray, double scale, Location result) {
-        int size = API.readMemoryInt(spriteArray, 0x40, 0x18);
+        long array_data = API.readMemoryLong(spriteArray, 0x40);
+        if (array_data == 0 ) return false;
+        int size = API.readMemoryInt(array_data + 0x20);
```
Comment (Pablete1234) [inline src/main/java/com/github/manolo8/darkbot/core/manager/MapManager.java:426]: these are equivalent except for the 0x18 -> 0x20 offset, readMemoryLong checks at every step of the way the ptr is valid now

File: src/main/java/com/github/manolo8/darkbot/core/manager/MapManager.java [hunk 3]
```diff
@@ -337,12 +337,12 @@ private void updateBounds() {
 
             renderValidated = false;
             is3DView = !main.settingsManager.is2DForced()
-                       && ByteUtils.readObjectName(API.readLong(viewAddress + 208)).contains("HUD");
+                && ByteUtils.readObjectName(API.readLong(viewAddress + 208)).contains("HUD");
```
Comment (Pablete1234) [inline src/main/java/com/github/manolo8/darkbot/core/manager/MapManager.java:340]: Undo the change plz

File: src/main/java/com/github/manolo8/darkbot/core/api/TanosAdapter.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+        private void sendNotification(String notification, int... args) {
+            if (botInstaller.screenManagerAddress.get() == 0) return;
+
+            long[] tagged = new long[args.length];
+            for (int i = 0; i < args.length; i++) {
+                tagged[i] = ByteUtils.tagInteger(args[i]);
+            }
+
+            tanos.sendNotification(botInstaller.screenManagerAddress.get(), notification, tagged);
+        }
```
Comment (Pablete1234) [inline src/main/java/com/github/manolo8/darkbot/core/api/TanosAdapter.java:142]: This whole piece should be moved to somewhere more reusable, potentially either GameApiImpl, or GameApi.DirectInteraction, where it expects another method (sendNotification) to be implemented and just calls that, that makes it so kekka and tanos don't need to copy-paste this block

File: src/main/java/com/github/manolo8/darkbot/core/manager/MouseManager.java [hunk 5]
```diff
@@ -1,6 +1,7 @@
 package com.github.manolo8.darkbot.core.manager;
 
 import com.github.manolo8.darkbot.Main;
+import com.github.manolo8.darkbot.core.entities.Box;
```
Comment (Pablete1234) [inline src/main/java/com/github/manolo8/darkbot/core/manager/MouseManager.java:4]:
```suggestion
```

File: src/main/java/com/github/manolo8/darkbot/core/utils/factory/EntityFactory.java [hunk 6]
```diff
@@ -132,8 +132,11 @@ public static EntityBuilder find(long address) {
     }
 
     private static String getAssetId(long address) {
-        long temp = API.readMemoryLong(address, 48, 48, 16) & ByteUtils.ATOM_MASK;
-        return API.readMemoryString(temp, 64, 32, 24, 8, 16, 24).trim();
+        // trait list
+        
+        long trait = API.readMemoryLong(address, 0x30, 0x30, 0x10) & ByteUtils.ATOM_MASK;
+        String r = API.readMemoryString(trait, 0x40, 0x20, 0x18, 0x10, 0x10, 0x18).trim();
+        return r;
```
Comment (Pablete1234) [inline src/main/java/com/github/manolo8/darkbot/core/utils/factory/EntityFactory.java:139]:
```suggestion
        long trait = API.readMemoryLong(address, 0x30, 0x30, 0x10) & ByteUtils.ATOM_MASK;
        return API.readMemoryString(trait, 0x40, 0x20, 0x18, 0x10, 0x10, 0x18).trim();
```

File: src/main/java/com/github/manolo8/darkbot/utils/ApiErrors.java [hunk 7]
```diff
@@ -26,7 +26,7 @@ private static String getMessage(BrowserApi api) {
             return "You're using a <strong>removed API</strong>, use the recommended API!";
         if (api == BrowserApi.NO_OP_API)
             return "Error loading the no-op api, how did this happen?";
-        if (api != BrowserApi.KEKKA_PLAYER)
+        if (api != BrowserApi.TANOS_API)
```
Comment (Pablete1234) [inline src/main/java/com/github/manolo8/darkbot/utils/ApiErrors.java:29]:
```suggestion
        if (api != (OSUtil.isLinux() ? BrowserApi.TANOS_API : BrowserApi.KEKKA_PLAYER))
```

maybe `OSUtil.getDefaultAPI` that just does this logic and re-use it in the 3 places it's used at?

**Review by Pablete1234 — CHANGES_REQUESTED** [review]
This is generally looking very, very compatible with the current codebase, unlike in the past where it was many, many changes.

I believe the main thing is sprite offset change from 0x18 to 0x20 in a few places, all other offset changes need to be double-checked but i believe they're just dec to hex conversions with the same value.

Overall very happy to see this be so close to a merge

**Review by Pablete1234 — APPROVED** [review]

**Pull request merged**